GPU/Software/Rasterizer.cpp

```diff
@@ -1119,86 +1109,139 @@ inline void DrawSinglePixel(const DrawingCoords &p, u16 z, u8 fog, const Vec4<in
+		GetTexelCoordinates(texlevel, s, t, u[0], v[0]);
+		if (frac_texlevel) {
+			GetTexelCoordinates(texlevel + 1, s, t, u[1], v[1]);
+		}
 
-		if (bilinear) {
-			GetTexelCoordinatesThroughQuad(texlevel, u_texel, v_texel, u, v);
-		} else {
-			GetTexelCoordinatesThrough(texlevel, u_texel, v_texel, u[0], v[0]);
+		texcolor0 = Vec4<int>::FromRGBA(SampleNearest<1>(texlevel, u, v, tptr0, bufwbytes0));
+		if (frac_texlevel) {
+			texcolor1 = Vec4<int>::FromRGBA(SampleNearest<1>(texlevel + 1, u + 1, v + 1, tptr1, bufwbytes1));
 		}
 	} else {
-		// we need to compute UV for a quad of pixels together in order to get the mipmap deltas :(
-		// texlevel = x
-
-		if (texlevel > maxTexLevel)
-			texlevel = maxTexLevel;
+		GetTexelCoordinatesQuad(texlevel, s, t, u, v, frac_u[0], frac_v[0]);
+		if (frac_texlevel) {
+			GetTexelCoordinatesQuad(texlevel + 1, s, t, u + 4, v + 4, frac_u[1], frac_v[1]);
+		}
 
-		if (bilinear) {
-			GetTexelCoordinatesQuad(texlevel, s, t, u, v, frac_u, frac_v);
-		} else {
-			GetTexelCoordinates(texlevel, s, t, u[0], v[0]);
+		texcolor0 = SampleLinear(texlevel, u, v, frac_u[0], frac_v[0], tptr0, bufwbytes0);
+		if (frac_texlevel) {
+			texcolor1 = SampleLinear(texlevel + 1, u + 4, v + 4, frac_u[1], frac_v[1], tptr1, bufwbytes1);
 		}
 	}
 
-	Vec4<int> texcolor;
-	int bufwbytes = texbufwidthbytes[texlevel];
-	const u8 *tptr = texptr[texlevel];
-	if (!bilinear) {
-		// Nearest filtering only. Round texcoords or just chop bits?
-		texcolor = Vec4<int>::FromRGBA(SampleNearest<1>(texlevel, u, v, tptr, bufwbytes));
-	} else {
-#if defined(_M_SSE)
-		Nearest4 c = SampleNearest<4>(texlevel, u, v, tptr, bufwbytes);
+	if (frac_texlevel) {
+		texcolor0 = (texcolor1 * frac_texlevel + texcolor0 * (256 - frac_texlevel)) / 256;
+	}
+	prim_color = GetTextureFunctionOutput(prim_color, texcolor0);
+}
 
-		const __m128i z = _mm_setzero_si128();
+static inline void ApplyTexturing(Vec4<int> *prim_color, const Vec4<float> &s, const Vec4<float> &t, int maxTexLevel, u8 *texptr[], int texbufwidthbytes[]) {
+	int width = gstate.getTextureWidth(0);
+	int height = gstate.getTextureHeight(0);
 
-		__m128i cvec = _mm_load_si128((const __m128i *)c.v);
-		__m128i tvec = _mm_unpacklo_epi8(cvec, z);
-		tvec = _mm_mullo_epi16(tvec, _mm_set1_epi16(0x100 - frac_v));
-		__m128i bvec = _mm_unpackhi_epi8(cvec, z);
-		bvec = _mm_mullo_epi16(bvec, _mm_set1_epi16(frac_v));
+	float ds = s[1] - s[0];
+	float dt = t[2] - t[0];
+	if (!gstate.isModeThrough()) {
+		ds *= width;
+		dt *= height;
+	}
 
-		// This multiplies the left and right sides.  We shift right after, although this may round down...
-		__m128i rowmult = _mm_set_epi16(frac_u, frac_u, frac_u, frac_u, 0x100 - frac_u, 0x100 - frac_u, 0x100 - frac_u, 0x100 - frac_u);
-		__m128i tmp = _mm_mulhi_epu16(_mm_add_epi16(tvec, bvec), rowmult);
+	float detail;
+	switch (gstate.getTexLevelMode()) {
+	case GE_TEXLEVEL_MODE_AUTO:
+		detail = std::log2f(std::abs(std::max(ds, dt)));
```

Comment:
optimization idea: instead of using the accurate log2f, pick apart the max(ds, dt) floating point number and use the exponent for the detail level and the mantissa as the interpolation factor towards the next one. I believe that's what most hardware implementations actually do.